changes.go

```diff
@@ -22,54 +22,294 @@ func FromData(data *charm.BundleData) []*Change {
+type Change interface {
+	// Id returns the unique identifier for this change.
+	Id() string
+	// Requires returns a list of dependencies for this change. Each dependency
 	// is represented by the corresponding change id, and must be applied
 	// before this change is applied.
-	Requires []string `json:"requires"`
+	Requires() []string
+	// Method returns the action to be performed to apply this change.
+	Method() string
+	// GUIArgs returns positional arguments to pass to the method, suitable for
+	// being serialized and sent to the Juju GUI.
+	GUIArgs() []interface{}
+	// setId is used to set the identifier for the change.
+	setId(string)
 }
 
-// changeset holds the list of changes returned by FromData.
-type changeset struct {
-	changes []*Change
+type changeInfo struct {
+	id       string
+	requires []string
+	method   string
+}
+
+// Id implements Change.Id.
+func (ch *changeInfo) Id() string {
+	return ch.id
+}
+
+// Requires implements Change.Requires.
+func (ch *changeInfo) Requires() []string {
+	if ch.requires == nil {
+		return []string{}
+	}
+	return ch.requires
+}
+
+// Method implements Change.Method.
+func (ch *changeInfo) Method() string {
+	return ch.method
+}
+
+// setId implements Change.setId.
+func (ch *changeInfo) setId(id string) {
+	ch.id = id
+}
+
+// newAddCharmChange creates a new change for adding a charm.
+func newAddCharmChange(args AddCharmArgs, requires ...string) *AddCharmChange {
+	return &AddCharmChange{
+		changeInfo: changeInfo{
+			requires: requires,
+			method:   "addCharm",
+		},
+		Args: args,
+	}
+}
+
+// AddCharmChange holds a change for adding a charm to the environment.
+type AddCharmChange struct {
+	changeInfo
+	// Args holds parameters for adding a charm.
+	Args AddCharmArgs
+}
+
+// GUIArgs implements Change.GUIArgs.
+func (ch *AddCharmChange) GUIArgs() []interface{} {
+	return []interface{}{ch.Args.Charm}
+}
+
+// AddCharmArgs holds parameters for adding a charm to the environment.
+type AddCharmArgs struct {
+	// Charm holds the URL of the charm to be added.
+	Charm string
+}
+
+// newAddMachineChange creates a new change for adding a machine or container.
+func newAddMachineChange(args AddMachineArgs, requires ...string) *AddMachineChange {
+	return &AddMachineChange{
+		changeInfo: changeInfo{
+			requires: requires,
+			method:   "addMachines",
+		},
+		Args: args,
+	}
+}
+
+// AddMachineChange holds a change for adding a machine or container.
+type AddMachineChange struct {
+	changeInfo
+	// Args holds parameters for adding a machine.
+	Args AddMachineArgs
+}
+
+// GUIArgs implements Change.GUIArgs.
+func (ch *AddMachineChange) GUIArgs() []interface{} {
+	options := AddMachineOptions{
+		Series:        ch.Args.Series,
+		Constraints:   ch.Args.Constraints,
+		ContainerType: ch.Args.ContainerType,
+		ParentId:      ch.Args.ParentId,
+	}
+	return []interface{}{options}
+}
+
+// AddMachineOptions holds GUI options for adding a machine or container.
```

Comment:
The type is used in tests too, I'll make it so that that's exported in some way, but I'd like to do that in another branch.